### PR TITLE
UpdateDuties improvement 1

### PR DESCRIPTION
**Pull request by mcdee**
**What type of PR is this?**

Bug fix

**What does this PR do? Why is it needed?**

This PR stops some non-attesting calculation from being in the critical path for attesting.

At current `UpdateDuties()` calculates attestation subnets and aggregation function for current and next epoch before returning.  It is possible for this to take a very long time if there is a high number of validators or slow CPU, due to the requirement to carry out two signatures per validator.  At worst case it goes on for multiple slots, meaning the validator is not validating when it should.

This PR separates the subnet and aggregation piece in to a separate goroutine, to move it out of the critical path.

**Which issues(s) does this PR fix?**

There are a number of issues raised that show poor attestation numbers.  This may or may not fix those individual issues, but it definitely fixes the situation where if the subnet calculation code takes too long to run it would result in missed proposals/attestations.

**Other notes for review**

This is part 1 of a three-part series.  The three parts are:

  1. move calculation for subnets out of the critical path
  2. optimise calculation for subnets
  3. cache 'next epoch' subnet results and reuse them if appropriate

### CODE REVIEW

**Comment by mcdee** [comment]
I wouldn't call it critical.  I looked at it this way:

Proposals:
  - if `UpdateDuties()` doesn't finish until after the end of the first slot, any proposal will be missed
  - if the goroutine doesn't finish until after the end of the first slot, no worries

Attestations:
  - if `UpdateDuties()` doesn't finish until after the end of the first slot, any attestation will be late, and possibly not included depending on how late
  - if the goroutine doesn't finish until after the end of the first slot, the attestation will still be made and broadcast but relies on there being either a pre-existing subnet or another node one hop away that is subscribed to the subnet for it to make its way to the relevant nodes for aggregation

The consideration for proposals is easy.  For attestations, I figure it is better to create the attestation and then potentially wait on the delivery of it than definitely wait on the creation (and even then have no guarantee that the delivery will be any better, given it can take time for the node to become a member of the subnet).

**Comment by terencechain** [comment]
> I wouldn't call it critical. I looked at it this way:
> 
> Proposals:
> 
> * if `UpdateDuties()` doesn't finish until after the end of the first slot, any proposal will be missed
> * if the goroutine doesn't finish until after the end of the first slot, no worries
> 
> Attestations:
> 
> * if `UpdateDuties()` doesn't finish until after the end of the first slot, any attestation will be late, and possibly not included depending on how late
> * if the goroutine doesn't finish until after the end of the first slot, the attestation will still be made and broadcast but relies on there being either a pre-existing subnet or another node one hop away that is subscribed to the subnet for it to make its way to the relevant nodes for aggregation
> 
> The consideration for proposals is easy. For attestations, I figure it is better to create the attestation and then potentially wait on the delivery of it than definitely wait on the creation (and even then have no guarantee that the delivery will be any better, given it can take time for the node to become a member of the subnet).

I agree with your reasoning. One last question, have you tested this change with attestant's validators?
This is something that can pass canary but fail run time in production

**Comment by mcdee** [comment]
Great question.  I've been running with a variant of this since day 2 of the new testnet, so I'm happy with the general idea.  What I haven't done is run the code in this PR in isolation from some other changes (pretty much part 2 of the above).  I'll set that up now and let it tick along for a day or so and report back.

**Comment by terencechain** [comment]
> Great question. I've been running with a variant of this since day 2 of the new testnet, so I'm happy with the general idea. What I haven't done is run the code in this PR in isolation from some other changes (pretty much part 2 of the above). I'll set that up now and let it tick along for a day or so and report back.

Sounds good, thanks! Ping us and we'll merge this

**Comment by terencechain** [comment]
I also worked on steaming validator duty, see: 
https://github.com/prysmaticlabs/prysm/issues/5306
https://hackmd.io/KWBj0PCDQUuPKpLQnIiZbw?both

Unfortunately I just haven't had the bandwidth to drive it to completion so anyone is more than welcome to jump in, just let me know which part so we dont duplicate work here

File: validator/client/validator.go [hunk 1]
```diff
@@ -330,73 +331,81 @@ func (v *validator) UpdateDuties(ctx context.Context, slot uint64) error {
...
-			subscribeIsAggregator = append(subscribeIsAggregator, aggregator)
 		}
-	}
 
-	_, err = v.validatorClient.SubscribeCommitteeSubnets(ctx, &ethpb.CommitteeSubnetsSubscribeRequest{
-		Slots:        subscribeSlots,
-		CommitteeIds: subscribeCommitteeIDs,
-		IsAggregator: subscribeIsAggregator,
-	})
+		_, err = v.validatorClient.SubscribeCommitteeSubnets(ctx, &ethpb.CommitteeSubnetsSubscribeRequest{
```
Comment (prestonvanloon) [inline validator/client/validator.go:401]: Please handle this error

File: validator/client/validator.go [hunk 2]
```diff
@@ -330,73 +331,81 @@ func (v *validator) UpdateDuties(ctx context.Context, slot uint64) error {
...
-	subscribeCommitteeIDs := make([]uint64, 0, len(validatingKeys))
-	subscribeIsAggregator := make([]bool, 0, len(validatingKeys))
-	alreadySubscribed := make(map[[64]byte]bool)
-
-	for _, duty := range v.duties.Duties {
-		if duty.Status == ethpb.ValidatorStatus_ACTIVE || duty.Status == ethpb.ValidatorStatus_EXITING {
-			attesterSlot := duty.AttesterSlot
-			committeeIndex := duty.CommitteeIndex
+	go func() {
+		// Create a new context as the existing one will be canceled when the parent function exits.
```
Comment (prestonvanloon) [inline validator/client/validator.go:335]: Can you break all of this out into another method?

File: validator/client/validator.go (hunk 1 above)
Comment (mcdee) [reply validator/client/validator.go:401]: Will do; thanks for the spot.

File: validator/client/validator.go (hunk 2 above)
Comment (mcdee) [reply validator/client/validator.go:335]: Can do; if I break it out simply the function call will look a little messy, _i.e._ `v.subscribeToSubnets(parentCtx, slot, validatingKeys, req)`  I was planning on a proper rewrite of this code to functions for the next part (with the removal of logging to a separate call there are two pieces of code that are virtually identical).  Three options:

  1. leave the code inline; deal with it in the next PR
  2. use the function call as outlined above; tidy up in the next PR
  3. bring forward the changes for the second phase in to this PR

Happy with any of these; let me know which you prefer.

Comment (mcdee) [reply validator/client/validator.go:335]: @prestonvanloon any thoughts on the above?

**Comment by nisdas** [comment]
@prestonvanloon can you take a look at this again ?

**Comment by rauljordan** [comment]
Hi @mcdee, we have modified our duties implementation to be a server-side stream https://github.com/prysmaticlabs/prysm/pull/5867, which the client listens to in the background. This new streaming approach solves the same problem as this PR, so we can close. Apologies for the long feedback cycle on this and thank you for the contribution nonetheless.

**Comment by rauljordan** [comment]
These are still crucial:
```
1. optimise calculation for subnets
2. cache 'next epoch' subnet results and reuse them if appropriate
```
we won't be touching the duties code after the stream duties functionality, so those additions would be very much welcome. For now, closing this PR

**Pull request closed without merging**